File: javascript/src/gameOfLife.test.js
import { lives } from './gameOfLife'

describe('Game of life', () => {
  const living = () => ({ isLiving: true })
  const dead = () => ({ isLiving: false })

  describe('Any live cell', () => {
    const cell = living()

    describe('with zero live neighbors', () => {
      const neighbors = [ dead(), dead(), dead(), dead() ]

      it('dies, as if by underpopulation', () => {
        expect(lives(cell, neighbors)).toEqual(false)
      })
    })
    describe('with one live neighbor', () => {
      const neighbors = [ living(), dead(), dead(), dead() ]

      xit('dies, as if by underpopulation', () => {
        expect(lives(cell, neighbors)).toEqual(false)
      })
    })
    describe('with two live neighbors', () => {
      const neighbors = [ living(), living(), dead(), dead() ]

      xit('lives on to the next generation.', () => {
        expect(lives(cell, neighbors)).toEqual(true)
      })
    })
    describe('with three live neighbors', () => {
      xit('lives on to the next generation.', () => {
      })
    })
    describe('with four live neighbors', () => {
      xit('dies, as if by overpopulation', () => {
      })
    })
  })

  describe('Any dead cell', () => {
    describe('with zero live neighbors', () => {
      xit('remains dead', () => {
      })
    })
    describe('with one live neighbor', () => {
      xit('remains dead', () => {
      })
    })
    describe('with two live neighbors', () => {
      xit('lives on to the next generation.', () => {
      })
    })
    describe('with three live neighbors', () => {
      xit('lives on to the next generation.', () => {
      })
    })
    describe('with four live neighbors', () => {
      xit('dies, as if by overpopulation', () => {
      })
    })
  })
})

File: python/tests/game_of_life_test.py
from game_of_life import lives, LIVING, DEAD

# Game of Life
class TestGameOfLife(object):

    # Any live cell
    class TestAnyLiveCell(object):

        # with zero live neighbours
        class TestZeroLiveNeighbours(object):

            # dies, as if by underpopulation
            def test_dies_by_underpopulation(self):
                CELL = LIVING
                NEIGHBOURS = [ DEAD, DEAD, DEAD, DEAD ]
                assert lives(CELL, NEIGHBOURS) == False

        # with one live neighbor
        class TestOneLiveNeighbours(object):

            # dies, as if by underpopulation
            def test_dies_by_underpopulation(self):
                CELL = LIVING
                NEIGHBOURS = [ LIVING, DEAD, DEAD, DEAD ]
                assert lives(CELL, NEIGHBOURS) == False

        # with two live neighbors
        class TestTwoLiveNeighbours(object):

            # lives on to the next generation.
            def test_lives_on_to_the_next_generation(self):
                CELL = LIVING
                NEIGHBOURS = [ LIVING, LIVING, DEAD, DEAD ]
                assert lives(CELL, NEIGHBOURS) == True

        # with three live neighbors
        class TestThreeLiveNeighbours(object):

            # lives on to the next generation.
            def test_lives_on_to_the_next_generation(self):
                CELL = LIVING
                NEIGHBOURS = [ LIVING, LIVING, LIVING, DEAD ]
                assert lives(CELL, NEIGHBOURS) == True

        # with four live neighbors
        class TestFourLiveNeighbours(object):

            # dies, as if by overpopulation
            def test_dies_by_underpopulation(self):
                CELL = LIVING
                NEIGHBOURS = [ LIVING, LIVING, LIVING, LIVING ]
                assert lives(CELL, NEIGHBOURS) == False

    # Any dead cell
        # with zero live neighbors
            # remains dead
        # with one live neighbor
            # remains dead
        # with two live neighbors
            # lives on to the next generation.
        # with three live neighbors
            # lives on to the next generation.
        # with four live neighbors
            # dies, as if by overpopulation
